README.md
# flutter_google_maps

![Build](https://github.com/marchdev-tk/flutter_google_maps/workflows/build/badge.svg)
[![Pub](https://img.shields.io/pub/v/flutter_google_maps.svg)](https://pub.dartlang.org/packages/flutter_google_maps)
![GitHub](https://img.shields.io/github/license/marchdev-tk/flutter_google_maps)
![GitHub stars](https://img.shields.io/github/stars/marchdev-tk/flutter_google_maps?style=social)

A Flutter plugin for integrating Google Maps in iOS, Android and Web applications. It is a wrapper of google_maps_flutter for Mobile and google_maps for Web.

## Getting Started

* Get an API key at <https://cloud.google.com/maps-platform/>.

* Enable Google Map SDK for each platform.
  * Go to [Google Developers Console](https://console.cloud.google.com/).
  * Choose the project that you want to enable Google Maps on.
  * Select the navigation menu and then select "Google Maps".
  * Select "APIs" under the Google Maps menu.
  * To enable Google Maps for Android, select "Maps SDK for Android" in the "Additional APIs" section, then select "ENABLE".
  * To enable Google Maps for iOS, select "Maps SDK for iOS" in the "Additional APIs" section, then select "ENABLE".
  * Make sure the APIs you enabled are under the "Enabled APIs" section.

* You can also find detailed steps to get start with Google Maps Platform [here](https://developers.google.com/maps/gmp-get-started).

### Web

```html
<body>
  <script src="https://maps.googleapis.com/maps/api/js?key=API_KEY"></script>
  <script src="main.dart.js" type="application/javascript"></script>
</body>
```

### Android

Specify your API key in the application manifest `android/app/src/main/AndroidManifest.xml`:

```xml
<manifest ...
  <application ...
    <meta-data android:name="com.google.android.geo.API_KEY"
               android:value="YOUR KEY HERE"/>
```

### iOS

Specify your API key in the application delegate `ios/Runner/AppDelegate.m`:

```objectivec
#include "AppDelegate.h"
#include "GeneratedPluginRegistrant.h"
#import "GoogleMaps/GoogleMaps.h"

@implementation AppDelegate

- (BOOL)application:(UIApplication *)application
    didFinishLaunchingWithOptions:(NSDictionary *)launchOptions {
  [GMSServices provideAPIKey:@"YOUR KEY HERE"];
  [GeneratedPluginRegistrant registerWithRegistry:self];
  return [super application:application didFinishLaunchingWithOptions:launchOptions];
}
@end
```

Or in your swift code, specify your API key in the application delegate `ios/Runner/AppDelegate.swift`:

```swift
import UIKit
import Flutter
import GoogleMaps

@UIApplicationMain
@objc class AppDelegate: FlutterAppDelegate {
  override func application(
    _ application: UIApplication,
    didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?
  ) -> Bool {
    GMSServices.provideAPIKey("YOUR KEY HERE")
    GeneratedPluginRegistrant.register(with: self)
    return super.application(application, didFinishLaunchingWithOptions: launchOptions)
  }
}
```
Opt-in to the embedded views preview by adding a boolean property to the app's `Info.plist` file
with the key `io.flutter.embedded_views_preview` and the value `YES`.

#### Android/iOS Directions API

Add in your `main.dart` within `main` function `GoogleMap.init('API_KEY');` before running the app.

```dart
void main() {
  GoogleMap.init('API_KEY');
  WidgetsFlutterBinding.ensureInitialized();
  runApp(MyApp());
}
```

For more info about **mobile** map setup, view [google_maps_flutter](https://pub.dev/packages/google_maps_flutter) plugin.

### Add GoogleMap Widget

```dart
import 'package:flutter/material.dart';
import 'package:flutter_google_maps/flutter_google_maps.dart';

...
GlobalKey<GoogleMapStateBase> _key = GlobalKey<GoogleMapStateBase>();

@override
Widget build(BuildContext context) => GoogleMap(
  key: _key,
),
...
```

And now you're ready to go.

## Examples

### GoogleMap widget can be configured with:

| Property          | Type                   | Description                                                                      |
| :---------------: | :--------------------: | :------------------------------------------------------------------------------: |
| initialPosition   | GeoCoord               | The initial position of the map's camera                                         |
| initialZoom       | double                 | The initial zoom of the map's camera                                             |
| mapType           | MapType                | Type of map tiles to be rendered                                                 |
| minZoom           | double                 | The preferred minimum zoom level or null, if unbounded from below                |
| maxZoom           | double                 | The preferred maximum zoom level or null, if unbounded from above                |
| mapStyle          | String                 | Sets the styling of the base map                                                 |
| mobilePreferences | MobileMapPreferences   | Set of mobile map preferences                                                    |
| webPreferences    | WebMapPreferences      | Set of web map preferences                                                       |
| interactive       | bool                   | Defines whether map is interactive or not                                        |
| onTap             | ValueChanged<GeoCoord> | Called every time a GoogleMap is tapped                                          |
| onLongPress       | ValueChanged<GeoCoord> | Called every time a GoogleMap is long pressed (for web when right mouse clicked) |
| markers           | Set<Marker>            | Markers to be placed on the map                                                  |

**`MapType` is one of following variants:**

  * `none` -> do not display map tiles
  * `roadmap` -> normal tiles (traffic and labels, subtle terrain information)
  * `satellite` -> satellite imaging tiles (aerial photos)
  * `terrain` -> terrain tiles (indicates type and height of terrain)
  * `hybrid` -> hybrid tiles (satellite images with some labels/overlays)

**`MobileMapPreferences` can be configured with:**

| Property                 | Type       | Description                                                                        |
| :----------------------: | :--------: | :--------------------------------------------------------------------------------: |
| compassEnabled           | bool       | True if the map should show a compass when rotated                                 |
| mapToolbarEnabled        | bool       | True if the map should show a toolbar when you interact with the map. Android only |
| myLocationEnabled        | bool       | True if a "My Location" layer should be shown on the map                           |
| myLocationButtonEnabled  | bool       | Enables or disables the my-location button                                         |
| indoorViewEnabled        | bool       | Enables or disables the indoor view from the map                                   |
| trafficEnabled           | bool       | Enables or disables the traffic layer of the map                                   |
| buildingsEnabled         | bool       | Enables or disables showing 3D buildings where available                           |
| padding                  | EdgeInsets | Padding to be set on mapdetails                                                    |
| rotateGesturesEnabled    | bool       | True if the map view should respond to rotate gestures                             |
| scrollGesturesEnabled    | bool       | True if the map view should respond to scroll gestures                             |
| zoomGesturesEnabled      | bool       | True if the map view should respond to zoom gestures                               |
| tiltGesturesEnabled      | bool       | True if the map view should respond to tilt gestures                               |

**`WebMapPreferences` can be configured with:**

| Property           | Type | Description                               |
| :----------------: | :--: | :---------------------------------------: |
| streetViewControl  | bool | Enables or disables streetViewControl     |
| fullscreenControl  | bool | Enables or disables fullscreenControl     |
| mapTypeControl     | bool | Enables or disables mapTypeControl        |
| scrollwheel        | bool | Enables or disables scrollwheel           |
| panControl         | bool | Enables or disables panControl            |
| overviewMapControl | bool | Enables or disables overviewMapControl    |
| rotateControl      | bool | Enables or disables rotateControl         |
| scaleControl       | bool | Enables or disables scaleControl          |
| zoomControl        | bool | Enables or disables zoomControl           |
| dragGestures       | bool | Enables or disables flutter drag gestures |

### To prepare for interacting with GoogleMap you will need to:

Create a `key` and assign it to the `GoogleMap` widget.

### GoogleMap widget has 2 static methods, they are:
  
* MapOperations of(GlobalKey<GoogleMapStateBase> key);
  
  Gets [MapOperations] interface via provided `key` of [GoogleMapStateBase] state.


* void init(String apiKey);
  
  Initializer of [GoogleMap]. `Required` if `Directions API` will be needed. For other cases, could be ignored.


### To interact with GoogleMap you'll need to:

**Use static `of` method**

Here's list of interactions:

* Move camera to the new bounds
  ```dart
  void moveCameraBounds(
    GeoCoordBounds newBounds, {
    double padding = 0,
    bool animated = true,
    bool waitUntilReady = true,
  });
  ```

* Move camera to the new coordinates
  ```dart
  void moveCamera(
    GeoCoord latLng, {
    bool animated = true,
    bool waitUntilReady = true,
    double zoom,
  });
  ```

* Zoom camera
  ```dart
  void zoomCamera(
    double zoom, {
    bool animated = true,
    bool waitUntilReady = true,
  });
  ```

* Get center coordinates of the map
  ```dart
  FutureOr<GeoCoord> get center;
  ```

* Change Map Style.

  The style string can be generated using [map style tool](https://mapstyle.withgoogle.com/).
  Also, refer [iOS](https://developers.google.com/maps/documentation/ios-sdk/style-reference)
  and [Android](https://developers.google.com/maps/documentation/android-sdk/style-reference)
  style reference for more information regarding the supported styles.

  ```dart
  void changeMapStyle(String mapStyle);
  ```

* Add marker to the map by given [position]
  ```dart
  void addMarkerRaw(
    GeoCoord position, {
    String label,
    String icon,
    String info,
    ValueChanged<String> onTap,
    VOidCallback onInfoWindowTap,
  });
  ```
**Please note:** [icon] could be a *path to an image asset* or it could be an instance of *ByteString*.

* Add marker to the map by given [marker] object
  ```dart
  void addMarker(Marker marker);
  ```

* Remove marker from the map by given [position]
  ```dart
  void removeMarker(GeoCoord position);
  ```

* Remove all markers from the map
  ```dart
  void clearMarkers();
  ```

* Add direction to the map by given [origin] and [destination] coordinates
  ```dart
  void addDirection(
    dynamic origin,
    dynamic destination, {
    String startLabel,
    String startIcon,
    String startInfo,
    String endLabel,
    String endIcon,
    String endInfo,
  });
  ```

* Remove direction from the map by given [origin] and [destination] coordinates
  ```dart
  void removeDirection(dynamic origin, dynamic destination);
  ```

* Remove all directions from the map
  ```dart
  void clearDirections();
  ```

* Add polygon to the map by given [id] and [points]
  ```dart
  void addPolygon(
    String id,
    Iterable<GeoCoord> points, {
    ValueChanged<String> onTap,
    Color strokeColor = const Color(0x000000),
    double strokeOpacity = 0.8,
    double strokeWidth = 1,
    Color fillColor = const Color(0x000000),
    double fillOpacity = 0.35,
  });
  ```

* Edit polygon on the map by given [id] and [points]
  ```dart
  void editPolygon(
    String id,
    Iterable<GeoCoord> points, {
    ValueChanged<String> onTap,
    Color strokeColor = const Color(0x000000),
    double strokeOpacity = 0.8,
    double strokeWeight = 1,
    Color fillColor = const Color(0x000000),
    double fillOpacity = 0.35,
  });
  ```

* Remove polygon from the map by given [id].
  ```dart
  void removePolygon(String id);
  ```

* Remove all polygones from the map.
  ```dart
  void clearPolygons();
  ```

## Feature requests and Bug reports

Feel free to post a feature requests or report a bug [here](https://github.com/marchdev-tk/flutter_google_maps/issues).

## TODO

* Add circles support
* Add polyline support


CHANGELOG.md
# Changelog

## [4.0.0]

* Added Circles support (thanks to [jan-pavlovsky](https://github.com/jan-pavlovsky))

## [3.8.0]

* Added moveCameraBounds (old moveCamera), moveCamera (with zooming ability) and zoomCamera (thanks to [travisjayday](https://github.com/travisjayday) for zoomControlsEnabled on mobile)
* Added center getter for center coordinates of the map
* Fixed issue when onTap/onLongPress was not specified

## [3.7.1]

* Fixed bug when app crash, when there's no any marker on mobile map (thanks to [Chojecki](https://github.com/Chojecki))

## [3.7.0]

* Changed onTap VoidCallback to onTap ValueChanged with marker ID as an argument
* Added onTap for polygon with polygon ID as an argument
* Added option for marker icon to be a path to an asset as well as to be a ByteString

### BREAKING CHANGE
* addMarker now is addMarkerRaw
* addMarker now accepts instance of a Marker object

## [3.6.0]

* Added infoSnippet as an argument to addMarker method

## [3.5.0]

* Fixed animateCamera issue and possible changeMapStyle issue

## [3.4.0]

* Added rotate/scroll/tilt/zoom gestures to mobile map preferences
* Added map callbacks onTap, onLongPress
* Added onInfoWindowTap for addMarker method on web

## [3.3.0]

* Added onInfoWindowTap for addMarker method (works only on Android/iOS)
* Added map iteractivity setup

## [3.2.0]

* Added onTap VoidCallback for addMarker method

## [3.1.0]

* Added gesture boolean to WebMapPreferences, that indicates whether to enable gestures or not

## [3.0.0]

* Implemented setting up Map Style

## [2.0.0]

* Fixed roadmap/normal Map Type

## [1.1.0]

* Update docs

## [1.0.0]

* Created wrapper for Mobile map and Web map
* Created common interface


lib/src/web/google_map.state.dart
// Copyright (c) 2020, the MarchDev Toolkit project authors. Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

import 'dart:html';
import 'dart:async';
import 'dart:ui' as ui;

import 'package:flutter/widgets.dart';
import 'package:flutter/scheduler.dart' show SchedulerBinding;

import 'package:uuid/uuid.dart';
import 'package:flinq/flinq.dart';
import 'package:google_maps/google_maps.dart';
import 'package:google_directions_api/google_directions_api.dart'
    show GeoCoord, GeoCoordBounds;

import 'utils.dart';
import '../core/google_map.dart';
import '../core/utils.dart' as utils;
import '../core/map_items.dart' as items;

class GoogleMapState extends GoogleMapStateBase {
  final htmlId = Uuid().v1();
  final directionsService = DirectionsService();

  final _markers = <String, Marker>{};
  final _infoState = <String, bool>{};
  final _infos = <String, InfoWindow>{};
  final _polygons = <String, Polygon>{};
  final _circles = <String, Circle>{};
  final _subscriptions = <StreamSubscription>[];
  final _directions = <String, DirectionsRenderer>{};

  GMap _map;
  MapOptions _mapOptions;

  String _getImage(String image) {
    if (image == null) return null;

    if (utils.ByteString.isByteString(image)) {
      final blob = Blob([utils.ByteString.fromString(image)], 'image/png');
      return Url.createObjectUrlFromBlob(blob);
    }

    return '${fixAssetPath(image)}assets/$image';
  }

  @override
  void moveCameraBounds(
    GeoCoordBounds newBounds, {
    double padding = 0,
    bool animated = true,
    bool waitUntilReady = true,
  }) {
    assert(() {
      if (newBounds == null) {
        throw ArgumentError.notNull('newBounds');
      }

      return true;
    }());

    _map.center = newBounds.center.toLatLng();

    final zoom = _map.zoom;
    if (animated == true) {
      _map.panToBounds(newBounds.toLatLngBounds());
    } else {
      _map.fitBounds(newBounds.toLatLngBounds());
    }
    _map.zoom = zoom;
  }

  @override
  void moveCamera(
    GeoCoord latLng, {
    bool animated = true,
    bool waitUntilReady = true,
    double zoom,
  }) {
    assert(() {
      if (latLng == null) {
        throw ArgumentError.notNull('latLng');
      }

      return true;
    }());

    if (animated == true) {
      _map.panTo(latLng.toLatLng());
      _map.zoom = zoom ?? _map.zoom;
    } else {
      _map.center = latLng.toLatLng();
      _map.zoom = zoom ?? _map.zoom;
    }
  }

  @override
  void zoomCamera(
    double zoom, {
    bool animated = true,
    bool waitUntilReady = true,
  }) {
    assert(() {
      if (zoom == null) {
        throw ArgumentError.notNull('zoom');
      }

      return true;
    }());

    _map.zoom = zoom;
  }

  @override
  FutureOr<GeoCoord> get center => _map.center?.toGeoCoord();

  @override
  void changeMapStyle(
    String mapStyle, {
    bool waitUntilReady = true,
  }) {
    try {
      _mapOptions.styles = mapStyle?.parseMapStyle();
      _map.options = _mapOptions;
    } catch (e) {
      throw utils.MapStyleException(e.toString());
    }
  }

  @override
  void addMarkerRaw(
    GeoCoord position, {
    String label,
    String icon,
    String info,
    String infoSnippet,
    ValueChanged<String> onTap,
    ui.VoidCallback onInfoWindowTap,
  }) {
    assert(() {
      if (position == null) {
        throw ArgumentError.notNull('position');
      }

      if (position.latitude == null || position.longitude == null) {
        throw ArgumentError.notNull('position.latitude && position.longitude');
      }

      return true;
    }());

    final key = position.toString();

    if (_markers.containsKey(key)) return;

    final marker = Marker()
      ..map = _map
      ..label = label
      ..icon = _getImage(icon)
      ..position = position.toLatLng();

    if (info != null || onTap != null) {
      _subscriptions.add(marker.onClick.listen((_) async {
        final key = position.toString();

        if (onTap != null) {
          onTap(key);
          return;
        }

        int doubleToInt(double value) => (value * 100000).truncate();
        final id =
            'position${doubleToInt(position.latitude)}${doubleToInt(position.longitude)}';

        if (_infos[key] == null) {
          print(id);
          final _info = onInfoWindowTap == null
              ? '$info${infoSnippet.isNotEmpty == true ? '\n$infoSnippet' : ''}'
              : '<p id="$id">$info${infoSnippet.isNotEmpty == true ? '<p>$infoSnippet</p>' : ''}</p>';

          _infos[key] = InfoWindow(InfoWindowOptions()..content = _info);
          _subscriptions.add(
              _infos[key].onCloseclick.listen((_) => _infoState[key] = false));
        }

        if (!(_infoState[key] ?? false)) {
          _infos[key].open(_map, marker);
          if (_infoState[key] == null) {
            await Future.delayed(const Duration(milliseconds: 100));

            final infoElem = querySelector('flt-platform-view')
                .shadowRoot
                .getElementById('$htmlId')
                .querySelector('#$id');

            infoElem.addEventListener('click', (event) => onInfoWindowTap());
          }
          _infoState[key] = true;
        } else {
          _infos[key].close();

          _infoState[key] = false;
        }
      }));
    }

    _markers[key] = marker;
  }

  @override
  void addMarker(items.Marker marker) => addMarkerRaw(
        marker.position,
        label: marker.label,
        icon: marker.icon,
        info: marker.info,
        infoSnippet: marker.infoSnippet,
        onTap: marker.onTap,
        onInfoWindowTap: marker.onInfoWindowTap,
      );

  @override
  void removeMarker(GeoCoord position) {
    assert(() {
      if (position == null) {
        throw ArgumentError.notNull('position');
      }

      if (position.latitude == null || position.longitude == null) {
        throw ArgumentError.notNull('position.latitude && position.longitude');
      }

      return true;
    }());

    final key = position.toString();

    var marker = _markers.remove(key);
    marker?.map = null;
    marker = null;

    var info = _infos.remove(key);
    info?.close();
    info = null;

    _infoState.remove(key);
  }

  @override
  void clearMarkers() {
    for (var marker in _markers.values) {
      marker?.map = null;
      marker = null;
    }
    _markers.clear();

    for (var info in _infos.values) {
      info?.close();
      info = null;
    }
    _infos.clear();

    _infoState.clear();
  }

  @override
  void addDirection(
    dynamic origin,
    dynamic destination, {
    String startLabel,
    String startIcon,
    String startInfo,
    String endLabel,
    String endIcon,
    String endInfo,
  }) {
    assert(() {
      if (origin == null) {
        throw ArgumentError.notNull('origin');
      }

      if (destination == null) {
        throw ArgumentError.notNull('destination');
      }

      return true;
    }());

    _directions.putIfAbsent(
      '${origin}_$destination',
      () {
        DirectionsRenderer direction = DirectionsRenderer(
            DirectionsRendererOptions()..suppressMarkers = true);
        direction.map = _map;

        final request = DirectionsRequest()
          ..origin = origin is GeoCoord
              ? LatLng(origin.latitude, origin.longitude)
              : origin
          ..destination =
              destination is GeoCoord ? destination.toLatLng() : destination
          ..travelMode = TravelMode.DRIVING;
        directionsService.route(
          request,
          (response, status) {
            if (status == DirectionsStatus.OK) {
              direction.directions = response;

              final leg = response?.routes?.firstOrNull?.legs?.firstOrNull;

              final startLatLng = leg?.startLocation;
              if (startLatLng != null) {
                if (startIcon != null ||
                    startInfo != null ||
                    startLabel != null) {
                  addMarkerRaw(
                    startLatLng.toGeoCoord(),
                    icon: startIcon,
                    info: startInfo ?? leg.startAddress,
                    label: startLabel,
                  );
                } else {
                  addMarkerRaw(
                    startLatLng.toGeoCoord(),
                    icon: 'assets/images/marker_a.png',
                    info: leg.startAddress,
                  );
                }
              }

              final endLatLng = leg?.endLocation;
              if (endLatLng != null) {
                if (endIcon != null || endInfo != null || endLabel != null) {
                  addMarkerRaw(
                    endLatLng.toGeoCoord(),
                    icon: endIcon,
                    info: endInfo ?? leg.endAddress,
                    label: endLabel,
                  );
                } else {
                  addMarkerRaw(
                    endLatLng.toGeoCoord(),
                    icon: 'assets/images/marker_b.png',
                    info: leg.endAddress,
                  );
                }
              }
            }
          },
        );

        return direction;
      },
    );
  }

  @override
  void removeDirection(dynamic origin, dynamic destination) {
    assert(() {
      if (origin == null) {
        throw ArgumentError.notNull('origin');
      }

      if (destination == null) {
        throw ArgumentError.notNull('destination');
      }

      return true;
    }());

    var value = _directions.remove('${origin}_$destination');
    value?.map = null;
    final start = value
        ?.directions?.routes?.firstOrNull?.legs?.firstOrNull?.startLocation
        ?.toGeoCoord();
    if (start != null) {
      removeMarker(start);
    }
    final end = value
        ?.directions?.routes?.firstOrNull?.legs?.lastOrNull?.endLocation
        ?.toGeoCoord();
    if (end != null) {
      removeMarker(end);
    }
    value = null;
  }

  @override
  void clearDirections() {
    for (var direction in _directions.values) {
      direction?.map = null;
      final start = direction
          ?.directions?.routes?.firstOrNull?.legs?.firstOrNull?.startLocation
          ?.toGeoCoord();
      if (start != null) {
        removeMarker(start);
      }
      final end = direction
          ?.directions?.routes?.firstOrNull?.legs?.lastOrNull?.endLocation
          ?.toGeoCoord();
      if (end != null) {
        removeMarker(end);
      }
      direction = null;
    }
    _directions.clear();
  }

  @override
  void addPolygon(
    String id,
    Iterable<GeoCoord> points, {
    ValueChanged<String> onTap,
    Color strokeColor = const Color(0x000000),
    double strokeOpacity = 0.8,
    double strokeWidth = 1,
    Color fillColor = const Color(0x000000),
    double fillOpacity = 0.35,
  }) {
    assert(() {
      if (id == null) {
        throw ArgumentError.notNull('id');
      }

      if (points == null) {
        throw ArgumentError.notNull('position');
      }

      if (points.isEmpty) {
        throw ArgumentError.value(<GeoCoord>[], 'points');
      }

      if (points.length < 3) {
        throw ArgumentError('Polygon must have at least 3 coordinates');
      }

      return true;
    }());

    _polygons.putIfAbsent(
      id,
      () {
        final options = PolygonOptions()
          ..clickable = onTap != null
          ..paths = points.mapList((_) => _.toLatLng())
          ..strokeColor = strokeColor?.toHashString() ?? '#000000'
          ..strokeOpacity = strokeOpacity ?? 0.8
          ..strokeWeight = strokeWidth ?? 1
          ..fillColor = strokeColor?.toHashString() ?? '#000000'
          ..fillOpacity = fillOpacity ?? 0.35;

        final polygon = Polygon(options)..map = _map;

        if (onTap != null) {
          _subscriptions.add(polygon.onClick.listen((_) => onTap(id)));
        }

        return polygon;
      },
    );
  }

  @override
  void editPolygon(
    String id,
    Iterable<GeoCoord> points, {
    ValueChanged<String> onTap,
    Color strokeColor = const Color(0x000000),
    double strokeOpacity = 0.8,
    double strokeWeight = 1,
    Color fillColor = const Color(0x000000),
    double fillOpacity = 0.35,
  }) {
    removePolygon(id);
    addPolygon(
      id,
      points,
      onTap: onTap,
      strokeColor: strokeColor,
      strokeOpacity: strokeOpacity,
      strokeWidth: strokeWeight,
      fillColor: fillColor,
      fillOpacity: fillOpacity,
    );
  }

  @override
  void removePolygon(String id) {
    assert(() {
      if (id == null) {
        throw ArgumentError.notNull('id');
      }

      return true;
    }());

    var value = _polygons.remove(id);
    value?.map = null;
    value = null;
  }

  @override
  void clearPolygons() {
    for (var polygon in _polygons.values) {
      polygon?.map = null;
      polygon = null;
    }
    _polygons.clear();
  }

  void _createMapOptions() {
    _mapOptions = MapOptions()
      ..zoom = widget.initialZoom
      ..center = widget.initialPosition.toLatLng()
      ..streetViewControl = widget.webPreferences.streetViewControl
      ..fullscreenControl = widget.webPreferences.fullscreenControl
      ..mapTypeControl = widget.webPreferences.mapTypeControl
      ..scrollwheel = widget.webPreferences.scrollwheel
      ..panControl = widget.webPreferences.panControl
      ..overviewMapControl = widget.webPreferences.overviewMapControl
      ..rotateControl = widget.webPreferences.rotateControl
      ..scaleControl = widget.webPreferences.scaleControl
      ..zoomControl = widget.webPreferences.zoomControl
      ..minZoom = widget.minZoom
      ..maxZoom = widget.maxZoom
      ..styles = widget.mapStyle?.parseMapStyle()
      ..mapTypeId = widget.mapType.toString().split('.')[1]
      ..gestureHandling = widget.interactive ? 'auto' : 'none';
  }

  @override
  void addCircle(
    String id,
    GeoCoord center,
    double radius, {
    ValueChanged<String> onTap,
    ui.Color strokeColor = const Color(0x000000),
    double strokeOpacity = 0.8,
    double strokeWidth = 1,
    ui.Color fillColor = const Color(0x000000),
    double fillOpacity = 0.35,
  }) {
    assert(() {
      if (id == null) {
        throw ArgumentError.notNull('id');
      }

      if (center == null) {
        throw ArgumentError.notNull('center');
      }

      if (radius == null) {
        throw ArgumentError.notNull('radius');
      }

      return true;
    }());

    _circles.putIfAbsent(
      id,
      () {
        final options = CircleOptions()
          ..center = center.toLatLng()
          ..radius = radius
          ..clickable = onTap != null
          ..strokeColor = strokeColor?.toHashString() ?? '#000000'
          ..strokeOpacity = strokeOpacity ?? 0.8
          ..strokeWeight = strokeWidth ?? 1
          ..fillColor = strokeColor?.toHashString() ?? '#000000'
          ..fillOpacity = fillOpacity ?? 0.35;

        final circle = Circle(options)..map = _map;

        if (onTap != null) {
          _subscriptions.add(circle.onClick.listen((_) => onTap(id)));
        }

        return circle;
      },
    );
  }

  @override
  void clearCircles() {
    for (var circle in _circles.values) {
      circle?.map = null;
      circle = null;
    }
    _circles.clear();
  }

  @override
  void editCircle(
    String id,
    GeoCoord center,
    double radius, {
    ValueChanged<String> onTap,
    ui.Color strokeColor = const Color(0x000000),
    double strokeOpacity = 0.8,
    double strokeWidth = 1,
    ui.Color fillColor = const Color(0x000000),
    double fillOpacity = 0.35,
  }) {
    removeCircle(id);
    addCircle(
      id,
      center,
      radius,
      onTap: onTap,
      strokeColor: strokeColor,
      strokeOpacity: strokeOpacity,
      strokeWidth: strokeWidth,
      fillColor: fillColor,
      fillOpacity: fillOpacity,
    );
  }

  @override
  void removeCircle(String id) {
    assert(() {
      if (id == null) {
        throw ArgumentError.notNull('id');
      }

      return true;
    }());

    var value = _circles.remove(id);
    value?.map = null;
    value = null;
  }

  @override
  void initState() {
    super.initState();
    SchedulerBinding.instance.addPostFrameCallback((_) {
      for (var marker in widget.markers) {
        addMarker(marker);
      }
    });
  }

  @override
  Widget build(BuildContext context) {
    _createMapOptions();

    if (_map == null) {
      ui.platformViewRegistry.registerViewFactory(htmlId, (int viewId) {
        final elem = DivElement()
          ..id = htmlId
          ..style.width = '100%'
          ..style.height = '100%'
          ..style.border = 'none';

        _map = GMap(elem, _mapOptions);

        _subscriptions.add(_map.onClick.listen(
            (event) => widget.onTap?.call(event?.latLng?.toGeoCoord())));
        _subscriptions.add(_map.onRightclick.listen(
            (event) => widget.onLongPress?.call(event?.latLng?.toGeoCoord())));

        return elem;
      });
    }

    return LayoutBuilder(
      builder: (context, constraints) => GestureDetector(
        onVerticalDragUpdate:
            widget.webPreferences.dragGestures ? null : (_) {},
        onHorizontalDragUpdate:
            widget.webPreferences.dragGestures ? null : (_) {},
        child: Container(
          constraints: BoxConstraints(maxHeight: constraints.maxHeight),
          child: HtmlElementView(viewType: htmlId),
        ),
      ),
    );
  }

  @override
  void dispose() {
    super.dispose();
    _subscriptions.forEach((_) => _.cancel());

    _infos.clear();
    _markers.clear();
    _polygons.clear();
    _circles.clear();
    _infoState.clear();
    _directions.clear();
    _subscriptions.clear();

    _map = null;
    _mapOptions = null;
  }
}
